File: five_hundred/five_hundred/player.py
from typing import List, Optional
from .card import Card, Suit, Rank

class Player:
    def __init__(self, name: str):
        self.name: str = name
        self.hand: List[Card] = []
        self.score: int = 0 # Individual score, team score might be managed elsewhere
        self.tricks_won_this_round: int = 0

    def add_card_to_hand(self, card: Card):
        """Adds a single card to the player's hand."""
        self.hand.append(card)

    def add_cards_to_hand(self, cards: List[Card]):
        """Adds multiple cards to the player's hand."""
        self.hand.extend(cards)

    def play_card(self, card_to_play: Card) -> Card:
        """Removes and returns a specific card from the player's hand."""
        if card_to_play not in self.hand:
            raise ValueError(f"Card {card_to_play} not in player's hand.")
        self.hand.remove(card_to_play)
        return card_to_play

    def sort_hand(self, trump_suit: Optional[Suit] = None, game_suit_order: Optional[List[Suit]] = None):
        """Sorts the player's hand. 
        Can be sorted by a specific suit order, then by rank (descending).
        If trump_suit is provided, trump cards come first, then other suits.
        Joker is usually the highest card.
        Note: Detailed card ranking (Bowers, Joker based on trump) will be complex 
        and might be better handled by a separate utility or within the Game logic 
        when determining playable cards or trick winners.
        For now, a simple sort by suit then rank.
        """
        # This is a placeholder for a more sophisticated sorting logic
        # that considers trump, bowers, joker value etc.
        # Standard sort order: Spades, Clubs, Diamonds, Hearts (example)
        # For now, just sort by suit enum order then by a predefined rank order.
        
        # Define a basic rank order (Ace high, Joker highest)
        # This will need to be dynamic based on trump suit in the actual game.
        rank_order = {rank: i for i, rank in enumerate(list(Rank))} # Ensure Ace is higher than King

        def sort_key(card: Card):
            suit_priority = list(Suit).index(card.suit)
            rank_priority = rank_order.get(card.rank, -1) # Joker might need special handling
            if card.rank == Rank.JOKER: # Make Joker highest
                return (-1, -1) # Arbitrary high priority
            return (suit_priority, -rank_priority) # -rank_priority for descending rank

        self.hand.sort(key=sort_key)


    def reset_for_new_round(self):
        """Resets hand and tricks won for a new round."""
        self.hand = []
        self.tricks_won_this_round = 0

    def increment_score(self, points: int):
        self.score += points

    def increment_tricks_won(self): 
        self.tricks_won_this_round += 1

    def __repr__(self):
        return f"Player({self.name}, Score: {self.score}, Hand: {len(self.hand)} cards, Tricks Won: {self.tricks_won_this_round})"

    def __str__(self):
        return f"{self.name} (Score: {self.score})" 

File: five_hundred_love/src/core/player.lua
local Utils = require "src.core.utils"
local CardModule = require "src.core.card"
local Suit = CardModule.Suit
local Rank = CardModule.Rank

local Player = Utils.class("Player")

function Player:init(name)
    self.name = name
    self.hand = {}
    self.score = 0
    self.tricks_won_this_round = 0
end

function Player:add_card_to_hand(card)
    table.insert(self.hand, card)
end

function Player:add_cards_to_hand(cards)
    for _, card in ipairs(cards) do
        table.insert(self.hand, card)
    end
end

function Player:play_card(card_to_play)
    -- Find and remove card
    for i, card in ipairs(self.hand) do
        if card == card_to_play then -- Uses Card.__eq if defined in metatable
            table.remove(self.hand, i)
            return card
        end
    end
    error("Card not in hand")
end

function Player:sort_hand()
    -- Sort by Suit (value) then Rank (value descending)
    -- Lua sort: returns true if a < b
    table.sort(self.hand, function(a, b)
        if a.suit ~= b.suit then
            return a.suit < b.suit
        else
            -- Joker check? 
            -- Rank.JOKER is 100, Ace is 14.
            -- If we want descending rank:
            return a.rank > b.rank
        end
    end)
end

function Player:reset_for_new_round()
    self.hand = {}
    self.tricks_won_this_round = 0
end

function Player:increment_score(points)
    self.score = self.score + points
end

function Player:increment_tricks_won()
    self.tricks_won_this_round = self.tricks_won_this_round + 1
end

function Player:__tostring()
    return string.format("%s (Score: %d)", self.name, self.score)
end

return Player
